### PR TITLE
TextQualityWatchdog Function Definition, Model Training Notebook, and Dataset Creation Scripts

### PR DESCRIPTION
**Pull request by WillChen789**
## Purpose

The changes include a new power skill, TextQualityWatchdog, that allows users to check the quality of text extracted during the document cracking process. Garbled/corrupt text will be detected by a pretrained machine learning model, and the corresponding document entry will be flagged in the index. Users can easily filter by the "text_quality_warning" field (0 for good quality, 1 for bad quality) to observe which documents could be problematic. This PR also includes utility scripts that can be used to build up a training dataset, as well as an ipynb demonstrating how to train, quantize, and evaluate the model. README.md is in progress.

## Does this introduce a breaking change?

```
[ ] Yes
[x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Get the code

```
git clone https://github.com/WillChen789/azure-search-power-skills.git
```

* Test the code

```
After cloning the repo, deploy the TextQualityWatchdog project as an Azure function app (https://docs.microsoft.com/en-us/azure/azure-functions/create-first-function-vs-code-python) to be tested as part of a skillset in Cognitive Search. See the "sample.dat" file for an example of the expected input, if testing locally in VS Code. The demo notebook is meant to be run sequentially cell-by-cell to show how the model was trained and evaluated.
```

## What to Check
Verify that the following are valid
* TrainWatchdogModel.ipynb notebook should be able to run on CPU

## Other Information

The training labels were computed from Tika OOV scores for each document, which means that:
* A value of 0 for "text_quality_warning" should indicate that the extracted text is high quality
* A value of 1 for "text_quality_warning" usually indicates that the extracted text is low quality, but note that this is correlated with OOV text that is still well-formed
* Therefore some false positives are expected

### CODE REVIEW

File: .vscode/extensions.json [hunk 1]
```diff
@@ -0,0 +1,6 @@
+{
```
Comment (bleroy) [inline .vscode/extensions.json:1]: Please gitignore your .vscode folder.

File: Text/.gitignore [hunk 2]
```diff
@@ -0,0 +1,130 @@
+# Byte-compiled / optimized / DLL files
```
Comment (bleroy) [inline Text/.gitignore:1]: You should only have a local gitignore if it needs to diverge significantly from the root one. If you have some reasonable changes to make to the root gitignore, that's fine.

File: Text/TextQualityWatchdog/Util/TikaExtract.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        # Get the full path to the metadata file
+        metadata_file_path = APP_ROOT_STR + "\\" + METADATA_DIR + "\\" + filename
+
+        # Download the blob to a local file
+        with open(download_file_path, "wb") as download_file:
+            download_file.write(blob_client.download_blob().readall())
+        print("Downloaded blob to \n\t" + download_file_path)
+
+        # Invoke tika server to perform extraction
+        parsedFile = parser.from_file(download_file_path)
```
Comment (bleroy) [inline Text/TextQualityWatchdog/Util/TikaExtract.py:100]: Given everything that can happen with Tika, can we add a little more around this to support timeouts, error handling, retries, etc.?

File: Text/TextQualityWatchdog/Util/TikaExtract.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+            metrics_dict['total'] += 1
+            metrics_dict['success'] += 1
+            with open(metrics_dict_file_path, 'wb') as picklefile:
+                pickle.dump(metrics_dict, picklefile)
+
+            print("Success!\n")
+
+        # Update high water mark
+        with open(last_read_file_path, "w") as last_read_file:
+            last_read_file.write(filename)
```
Comment (bleroy) [inline Text/TextQualityWatchdog/Util/TikaExtract.py:154]: Would it be useful to put a write lock on that file while the process is running?

File: Text/TextQualityWatchdog/Util/TikaExtract.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+        # Iterate until we hit the high water mark
+        if last_read_file_name != None and filename < last_read_file_name:
+            print("skip!")
+            continue
+
+        # Get the blob client for the current blob
+        blob_client = cont_client.get_blob_client(blobname)
+
+        # Get the path to save the unstructured file to
+        download_file_path = APP_ROOT_STR + "\\" + DOWNLOAD_DIR + "\\" + filename
```
Comment (bleroy) [inline Text/TextQualityWatchdog/Util/TikaExtract.py:86]: Windows-only. Please use the join API like you did in other files.

File: Text/TextQualityWatchdog/__init__.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+        if body:
+            logging.info(body)
+            values = json.loads(body)['values']
+            results = {}
+            results["values"] = []
+
+            for value in values:
+                text = value['data']['text']
+
+                # Apply puntuation and whitespace normalization, and conert to lowercase
```
Comment (bleroy) [inline Text/TextQualityWatchdog/__init__.py:43]: supernit: typo on convert

File: Text/TextQualityWatchdog/__init__.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+                # Compute the input tokens and attention masks for the text sequence
+                input_ids, attention_masks = get_ids_and_masks(tokenizer, text)
+
+                # Call the ONNX model to perform inference on the input
+                flat_prediction = predict(ort_session, input_ids, attention_masks)
+
+                payload = ({"recordId": value['recordId'],   
+                            "data": {"text_quality_warning": int(flat_prediction[0])}
+                            })
```
Comment (bleroy) [inline Text/TextQualityWatchdog/__init__.py:57]: Nit: spread on more lines for better readability.

**Review by bleroy — CHANGES_REQUESTED** [review]
Thanks for the contribution. You already know that, but please add readme and deployment instructions consistently with other power skills, and update the top-level readme to point here.

File: Text/TextQualityWatchdog/Util/NormalizeData.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+    return corrected
+
+def _normalize_whitespace(text):
+    """
+    This function normalizes whitespaces, removing duplicates.
+    """
+    corrected = str(text)
+    corrected = re.sub('\s', ' ', corrected)
+    corrected = re.sub(r"( )\1+",r"\1", corrected)
+    return corrected.strip(" ")
```
Comment (jadrefke) [inline Text/TextQualityWatchdog/Util/NormalizeData.py:39]: Just inmport PreProcessData and reuse these methods from there (or visa versa). Best not to dupe code if at all possible

**Comment by jadrefke** [comment]
>         "quantized_model = quantize(\r\n",

Warning: onnxruntime.quantization.quantize is deprecated.
         Please use quantize_static for static quantization, quantize_dynamic for dynamic quantization.

---
Refers to: Text/TextQualityWatchdog/Model/TrainWatchdogModel.ipynb:1297 in 55f7e07. [](commit_id = 55f7e07b89cb0178d7f132c621a97027ffd2922a, deletion_comment = False)

**Comment by jadrefke** [comment]
>         "export(\r\n",

Warning: The original model opset version is 9, which does not support quantization. Please update the model to opset >= 11. Updating the model automatically to opset 11. Please verify the quantized model.

---
Refers to: Text/TextQualityWatchdog/Model/TrainWatchdogModel.ipynb:1243 in 55f7e07. [](commit_id = 55f7e07b89cb0178d7f132c621a97027ffd2922a, deletion_comment = False)

File: Text/TextQualityWatchdog/README.md [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+When included in an index, this value can be used to filter and examine documents that had low quality text extracted during document cracking.
+
+## Requirements
+
+This function has no additional requirements outside the common requirements described in [the root `README.md` file](../../README.md).
+
+## Optional Model Customization
+
+Please note that the provided "watchdog_model.onnx" classifier was trained on a dataset of 18,000 documents, 50% of which were synthetically generated to correct for class imbalance.
+Additionally, labels were generated using an "out-of-vocabulary" metric, which means that a "text_quality_warning" value of 1 does not necessarily indicate corrupted/garbled text.
```
Comment (jadrefke) [inline Text/TextQualityWatchdog/README.md:27]: > does not necessarily indicate corrupted/garbled text. [](http://example.com/codeflow?start=125&length=54)

nit: would phrase this "means that the data isn't using comon words, which often correlated with garbled text"

Comment (WillChen789) [reply Text/TextQualityWatchdog/README.md]: Done! Rephrased this section about the OOV score in README.md

File: Text/TextQualityWatchdog/Util/NormalizeData.py (hunk 8 above)
Comment (WillChen789) [reply Text/TextQualityWatchdog/Util/NormalizeData.py]: Done! Removed redundant methods and added import statement.

File: Text/TextQualityWatchdog/__init__.py (hunk 7 above)
Comment (WillChen789) [reply Text/TextQualityWatchdog/__init__.py:57]: Thanks for the advice! Fixed!

File: Text/TextQualityWatchdog/__init__.py (hunk 6 above)
Comment (WillChen789) [reply Text/TextQualityWatchdog/__init__.py:43]: Fixed!

File: Text/TextQualityWatchdog/Util/TikaExtract.py [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+# Process each blob
+for cur_blob in blob_list:
+    try:
+        # Get the name of the current blob
+        blobname = cur_blob.name
+        filename = blobname.split('/')[1]
+        print(blobname)
+
+        # Iterate until we hit the high water mark
+        if last_read_file_name != None and filename < last_read_file_name:
```
[the remark this answers is not in the document]
Comment (WillChen789) [reply Text/TextQualityWatchdog/Util/TikaExtract.py:78]: That's a great point, Matt! I designed this tool to specifically work with our CommonCrawl dataset, which is static. Since this isn't meant to be a general-purpose tool, it's safe to keep track of the last read file for our purposes. I've added a comment in the file header to be more clear about the use case.

File: Text/TextQualityWatchdog/Util/TikaExtract.py [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+            # Update metrics and save to disk
+            metrics_dict['total'] += 1
+            metrics_dict['success'] += 1
+            with open(metrics_dict_file_path, 'wb') as picklefile:
+                pickle.dump(metrics_dict, picklefile)
+
+            print("Success!\n")
+
+        # Update high water mark
+        with open(last_read_file_path, "w") as last_read_file:
```
[the remark this answers is not in the document]
Comment (WillChen789) [reply Text/TextQualityWatchdog/Util/TikaExtract.py:153]: I believe that the first update to the high water mark is conditional: if Tika fails to extract text from a document, we still want to update the high water mark and then move on to the next document.

File: Text/TextQualityWatchdog/Util/TikaExtract.py [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+for cur_blob in blob_list:
+    try:
+        # Get the name of the current blob
+        blobname = cur_blob.name
+        filename = blobname.split('/')[1]
+        print(blobname)
+
+        # Iterate until we hit the high water mark
+        if last_read_file_name != None and filename < last_read_file_name:
+            print("skip!")
```
[the remark this answers is not in the document]
Comment (WillChen789) [reply Text/TextQualityWatchdog/Util/TikaExtract.py]: Thanks Matt! That's a good point since we've got a lot of files. I added a verbose flag so that users can selectively print out logging output.

File: .vscode/extensions.json (hunk 1 above)
Comment (WillChen789) [reply .vscode/extensions.json:1]: Added .vscode to root gitignore in /Text

File: Text/.gitignore (hunk 2 above)
Comment (WillChen789) [reply Text/.gitignore:1]: Added .vscode to root gitignore in /Text

File: Text/TextQualityWatchdog/Util/TikaExtract.py (hunk 4 above)
Comment (WillChen789) [reply Text/TextQualityWatchdog/Util/TikaExtract.py:171]: Thanks for the suggestion Bertrand! I wrote this helper script with single-threading in mind, and I believe that the "with open" statement is conventional in Python since it creates a context manager to handle file I/O correctly.

File: Text/TextQualityWatchdog/Util/TikaExtract.py (hunk 5 above)
Comment (WillChen789) [reply Text/TextQualityWatchdog/Util/TikaExtract.py]: That's good to know! I made the changes to use the join API, and tested the code again to verify that the behavior hasn't changed.

File: Text/TextQualityWatchdog/Util/TikaExtract.py (hunk 3 above)
Comment (WillChen789) [reply Text/TextQualityWatchdog/Util/TikaExtract.py]: That's a good call. From my limited experience with Tika, I've found that a lot of the time, simply rerunning the parser will resolve the errors that it had the first time. So, I added a PARSER_MAX_RETRIES constant to control how many times to retry the Tika parser if it fails to extract text, as well as more explicit exception handling logic.

File: .vscode/extensions.json (hunk 1 above)
Comment (jadrefke) [reply .vscode/extensions.json:1]: you also need to remove hte files from your commit

**Review by jadrefke — APPROVED** [review]
:shipit:
